TESTING.md

```diff
@@ -0,0 +1,358 @@
+
+## Table of Contents
+
+- [Prerequisites](#prerequisites)
+- [Test Types](#test-types)
+- [Running Tests](#running-tests)
+- [Test Configuration](#test-configuration)
+- [Development Testing](#development-testing)
+- [Continuous Integration](#continuous-integration)
+- [Troubleshooting](#troubleshooting)
+
+## Prerequisites
+
+### Required Software
+
+1. **Java 11 or higher** (OpenJDK or Oracle)
+2. **Docker** for running integration tests
+3. **Git** for cloning repositories
+
+### Build Prerequisites
+
+Before running tests, you must build the required Cassandra dtest jars:
+
+```bash
+./scripts/build-dtest-jars.sh
+```
+
+This script builds dtest jars for the supported Cassandra versions.
+You can customize which versions to build:
+
+- **REPO**: Cassandra git repository (default: `https://github.com/apache/cassandra.git`)
+- **BRANCHES**: Space-delimited list of branches (default: `"cassandra-4.0 cassandra-4.1 cassandra-5.0 trunk"`)
+
+Example with custom branches:
+```bash
+BRANCHES="cassandra-4.1 trunk" ./scripts/build-dtest-jars.sh
+```
+
+### Network Setup for Multi-Node Tests
+
+For multi-node in-jvm dtests, network aliases must be configured.
+Tests assume each node's IP address is `127.0.0.x`, where `x` is the node ID.
+
+#### macOS Network Aliases
+
+Create temporary aliases for nodes 2-20:
+```bash
+for i in {2..20}; do sudo ifconfig lo0 alias "127.0.0.${i}"; done
+```
+
+#### Linux Network Aliases
+
+Create temporary aliases for nodes 2-20:
+```bash
+for i in {2..20}; do sudo ip addr add "127.0.0.${i}/8" dev lo; done
+```
+
+## Test Types
+
+### Unit Tests
+
+Unit tests verify individual components in isolation. They are fast-running
+and don't require external dependencies.
+
+**Location**: `src/test/java/` in each module
+
+**Running unit tests**:
+```bash
+# Run all unit tests
+./gradlew test
+
+# Run unit tests for a specific module
+./gradlew :client:test
+./gradlew :server:test
+```
+
+### Integration Tests
+
+Integration tests verify component interactions and require running
+Cassandra instances. They are divided into two categories:
+
+#### Lightweight Integration Tests
+- Basic functionality tests
+- Faster execution
+- **Tag**: None (default)
+
+#### Heavyweight Integration Tests  
+- Complex scenarios
+- Multi-node cluster tests
+- Longer execution times
+- **Tag**: `@Tag("heavy")`
+
+**Location**: `src/integrationTest/` in test modules
+
+**Running integration tests**:
+```bash
+# Run all integration tests
+./gradlew integrationTest
+
+# Run only lightweight integration tests
+./gradlew integrationTestLightWeight
+
+# Run only heavyweight integration tests
+./gradlew integrationTestHeavyWeight
+
+# Run integration tests for specific module
+./gradlew :integration-tests:integrationTest
+
+# Run integration tests for specific module
+./gradlew :server:integrationTest
+```
+
+### Container Tests
+
+Container-based tests run in Docker environments for additional isolation.
```

Comment:
We should probably describe the cases where we use container tests (just the SBW-on-S3 code where we need the S3 mock, right?) so folks don't unintentionally add tests where they aren't needed.